### PR TITLE
txn: use tokio_threadpool for txn scheduler

### PR DESCRIPTION
**Pull request by fredchenbj**
Signed-off-by: fredchenbj <cfworking@163.com>

## What have you changed? (mandatory)
use tokio-threadpool to replace threadpool in tikv_util used by `txn` scheduler. 

The currently defined types are listed below, please pick one of the types for this PR by removing the others:
- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
manual tests

### CODE REVIEW

**Comment by siddontang** [comment]
Great, PTAL @breeswish @hicqu

**Comment by siddontang** [comment]
@fredchenbj 
can you show us the benchmark results?

**Comment by zhangjinpeng1987** [comment]
@fredchenbj Do you test the binary with grafana and check if these metrics work well?

**Comment by fredchenbj** [comment]
I use `go-ycsb` tool with command: `./bin/go-ycsb load/run tikv -P workloads/workloada -p tikv.pd="10.136.16.2:2379" -p tikv.type="txn" -p tikv.conncount=32 --threads=32 --target=3000`. The result of test case is followed:
```
load
master: Avg(us): 6779, Min(us): 4925, Max(us): 393967, 95th(us): 9000, 99th(us): 13000

tokio:  Avg(us): 6608, Min(us): 4743, Max(us): 363260, 95th(us): 8000, 99th(us): 13000

run
master: Avg(us): 2748, Min(us): 1898, Max(us): 229763, 95th(us): 4000, 99th(us): 5000

tokio:  Avg(us): 2841, Min(us): 1929, Max(us): 210304, 95th(us): 4000, 99th(us): 5000
```
From above, the latency under certain load is not decreased obviously, and the context switch decreased 5%. 
@siddontang

**Comment by fredchenbj** [comment]
> @fredchenbj Do you test the binary with grafana and check if these metrics work well?

I have tested the binary, I would check metrics more carefully and comment here later.

**Comment by fredchenbj** [comment]
> @fredchenbj Do you test the binary with grafana and check if these metrics work well?

I have tested the binary, I would check metrics more carefully and comment here later.

**Comment by zhangjinpeng1987** [comment]
> and the context switch decreased about 5%

@fredchenbj I wonder why the context switch decreased, do you have any idea?

File: src/storage/txn/sched_pool_impl.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+pub fn build_sched_pool(pool_size: usize, name_prefix: &str) -> FuturePool {
+    FuturePoolBuilder::new()
+        .pool_size(pool_size)
+        .name_prefix(name_prefix)
+        .on_tick(move || tls_flush())
+        .before_stop(move || tls_flush())
+        .build()
+}
+
+#[inline]
```
Comment (ngaut) [inline src/storage/txn/sched_pool_impl.rs]: Please remove all of the inline hints, let the compiler decide to optimize the code.

Comment (ngaut) [reply src/storage/txn/sched_pool_impl.rs]: I know, we don't need to do that, usually the compiler is smart enough to make the decision.

**Comment by fredchenbj** [comment]
> @fredchenbj Do you test the binary with grafana and check if these metrics work well?

I have check these metrics work as before. 

> > and the context switch decreased about 5%
> 
> @fredchenbj I wonder why the context switch decreased, do you have any idea?

I don't have any ideas about this.

File: src/storage/txn/sched_pool.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            .pool_size(pool_size)
+            .name_prefix(name_prefix)
+            .on_tick(move || tls_flush())
+            .before_stop(move || tls_flush())
+            .build();
+        SchedPool { engine, pool }
+    }
+}
+
+pub fn build_sched_pool(pool_size: usize, name_prefix: &str) -> FuturePool {
```
Comment (breeswish) [inline src/storage/txn/sched_pool.rs]: The name is quite confusing.. `build_sched_pool` does not build a `SchedPool` but a `FuturePool` instead.

File: src/storage/metrics.rs [hunk 3]
```diff
@@ -5,6 +5,9 @@ use prometheus_static_metric::*;
 
 make_static_metric! {
     pub label_enum CommandKind {
+        get,
```
Comment (breeswish) [inline src/storage/metrics.rs]: Looks like this change is not necessary?

File: src/storage/txn/process.rs [hunk 4]
```diff
@@ -111,28 +108,25 @@ impl Task {
...
 impl<E: Engine> Executor<E> {
-    pub fn new(
-        scheduler: worker::Scheduler<Msg>,
-        pool: threadpool::Scheduler<SchedContext<E>>,
-    ) -> Self {
+    pub fn new(scheduler: worker::Scheduler<Msg>, pool: SchedPool<E>) -> Self {
         Executor {
             scheduler: Some(scheduler),
-            pool: Some(pool),
+            schedpool: Some(pool),
```
Comment (breeswish) [inline src/storage/txn/process.rs]: prefer `sched_pool`

**Comment by breeswish** [comment]
@zhangjinpeng1987 A work-stealing pool have local task queues, which may help with reducing context switch.

**Comment by breeswish** [comment]
@zhangjinpeng1987 A work-stealing pool have local task queues, which may help with reducing context switch.

File: src/storage/metrics.rs (hunk 3 above)
Comment (fredchenbj) [reply src/storage/metrics.rs]: This change will make metrics `tikv_storage_command_total` have type `get` , then `command` would have this type. I tested that without this change, `tikv_storage_command_total` doesn't have type `get`.

Comment (fredchenbj) [reply src/storage/metrics.rs]: yes, here I tested wrong before, and will remove it.

**Comment by fredchenbj** [comment]
friendly ping @siddontang @breeswish

File: src/storage/txn/process.rs [hunk 5]
```diff
@@ -181,50 +175,46 @@ impl<E: Engine> Executor<E> {
...
         let readonly = task.cmd.readonly();
-        pool.schedule(move |ctx: &mut SchedContext<E>| {
+        sched_pool.pool.spawn(move || {
             fail_point!("scheduler_async_snapshot_finish");
 
-            let _processing_read_timer = ctx
-                .processing_read_duration
-                .with_label_values(&[tag])
-                .start_coarse_timer();
+            let read_duration = tikv_util::time::Instant::now_coarse();
```
Comment (siddontang) [inline src/storage/txn/process.rs]: can we use `Instant::now_coarse()` here?

Comment (fredchenbj) [reply src/storage/txn/process.rs]: fixed it.

**Comment by siddontang** [comment]
PTAL @hicqu 

please fix the confliect @fredchenbj

**Comment by hicqu** [comment]
I merged master and resolved some conflicts. PTAL again @breeswish thanks!

**Comment by hicqu** [comment]
LGTM

**Pull request merged**